### PR TITLE
Testing another approach to adding clip and scale to throttle curve

### PR DESCRIPTION
**Pull request by ChrisRosser**
Not ready to merge. For testing only. Complementary to #3871

### CODE REVIEW

File: src/js/tabs/pid_tuning.js [hunk 1]
```diff
@@ -1758,6 +1780,27 @@ pid_tuning.initialize = function (callback) {
...
+                const clippos = throttleLimitPercent <= midy
+                    ? getPosfromYBezier(throttleLimitPercent,canvasHeight,midyl, midy)
+                    : getPosfromYBezier(throttleLimitPercent,midy, midyl, 0);
+                if (throttleLimitPercent <= midyl){
+                    //first half
+                    lowerctrl = getQuadraticCurvePoint(0, canvasHeight, midxl, midyl, midx, midy, clippos/2);
+                    curveclip = getQuadraticCurvePoint(0, canvasHeight, midxl, midyl, midx, midy, clippos);
+                    context.quadraticCurveTo(lowerctrl.x, lowerctrl.y, curveclip.x, curveclip.y);
+                    context.moveTo(curveclip.x, curveclip.y);
+                    context.lineTo(canvasWidth, curveclip.y);
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1793]: You can put these lines after the if else block - as they are the same in both conditions

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1793]: Same for lines 1799 - 1801.

Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1793]: Lines 1799 - 1801 won't be needed.

File: src/js/tabs/pid_tuning.js [hunk 2]
```diff
@@ -1715,13 +1715,34 @@ pid_tuning.initialize = function (callback) {
...
+            Maths from: https://stackoverflow.com/questions/40918569/quadratic-bezier-curve-calculate-x-for-any-given-y
+            */
+            function getPosfromYBezier(y, startY, cpY, endY){
+            // y = (1-t)^2 * p0 + 2 * (1-t)*t*p1 + t^2 * p2
+            // y = (p2+p0-2p1)x^2 + 2(p1 - p0)x + p0
+            // 0 = (p2+p0-2p1)x^2 + 2(p1 - p0)x + (p0 - y)
+            const a = startY+endY-2*cpY;
+            const b = 2*(cpY - startY);
+            const c = startY - y;
+            return a==0? -c/b : (-b + Math.sqrt(Math.pow(b, 2) - 4 * a * c)) / (2 * a);
```
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js:1728]: image

[the remark this answers is not in the document]
Comment (ChrisRosser) [reply src/js/tabs/pid_tuning.js]: I think I've fixed those errors and made the indentation and formatting changes

File: src/js/tabs/pid_tuning.js [hunk 3]
```diff
@@ -1758,9 +1780,26 @@ pid_tuning.initialize = function (callback) {
             context.clearRect(0, 0, canvasWidth, canvasHeight);
             context.beginPath();
             context.moveTo(0, canvasHeight);
+            if (throttleLimitType === THROTTLE_LIMIT_TYPES.CLIP){
+                const clippos = throttleLimitPercent <= midy
+                    ? getPosfromYBezier(throttleLimitPercent,canvasHeight,midyl, midy)
+                    : getPosfromYBezier(throttleLimitPercent,midy, midyl, 0);
+                let lowerctrl = getQuadraticCurvePoint(0, canvasHeight, midxl, midyl, midx, midy, clippos/2);
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1787]:
```suggestion
                let lowerctrl = getQuadraticCurvePoint(0, canvasHeight, midxl, midyl, midx, midy, clippos / 2);
```

File: src/js/tabs/pid_tuning.js [hunk 4]
```diff
@@ -1758,9 +1780,26 @@ pid_tuning.initialize = function (callback) {
...
+                let curveclip = getQuadraticCurvePoint(0, canvasHeight, midxl, midyl, midx, midy, clippos);
+                if (throttleLimitPercent > midyl){
+                    context.quadraticCurveTo(midxl, midyl, midx, midy);
+                    context.moveTo(midx, midy);
+                    lowerctrl = getQuadraticCurvePoint(midx, midy, midxr, midyr, canvasWidth, 0, clippos/2);
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1792]:
```suggestion
                    lowerctrl = getQuadraticCurvePoint(midx, midy, midxr, midyr, canvasWidth, 0, clippos / 2);
```

File: src/js/tabs/pid_tuning.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: Sorry we also need to have the variables in camelCase :joy:

[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: And one left :see_no_evil: 

image

[the remark this answers is not in the document]
Comment (ChrisRosser) [reply src/js/tabs/pid_tuning.js]: @haslinghuis Needs a bit more work. I'll ping you again when I think it's ready!

[the remark this answers is not in the document]
Comment (haslinghuis) [reply src/js/tabs/pid_tuning.js]: No worries - just highlighting. Take time needed.

File: src/js/tabs/pid_tuning.js [hunk 5]
```diff
@@ -1738,29 +1759,48 @@ pid_tuning.initialize = function (callback) {
...
+                    curveClip = getQuadraticCurvePoint(midx, midy, midxr, midyr, canvasWidth, topy, clipPos);
+                }
+                context.quadraticCurveTo(lowerCtrl.x, lowerCtrl.y, curveClip.x, curveClip.y);
+                context.moveTo(curveClip.x, curveClip.y);
+                context.lineTo(canvasWidth, curveClip.y);
+            } else {
             context.quadraticCurveTo(midxl, midyl, midx, midy);
             context.moveTo(midx, midy);
-            context.quadraticCurveTo(midxr, midyr, canvasWidth, 0);
+            context.quadraticCurveTo(midxr, midyr, canvasWidth, topy);
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:1802]: :space_invader:

**Comment by ChrisRosser** [comment]
@haslinghuis I think the maths is now correct but I'm getting an issue with the curve not plotting in CLIP mode. Any help would be much appreciated.

**Pull request closed without merging**